README.md
# Tutorial

<img src="https://user-images.githubusercontent.com/47983071/105987554-30276a00-607d-11eb-9f0f-69acfad9e8f8.jpeg" alt="drawing" width="200"/>

<img src="https://user-images.githubusercontent.com/47983071/105987620-4cc3a200-607d-11eb-958c-a5645bba87fb.jpeg" alt="drawing" width="200"/>


<img src="https://user-images.githubusercontent.com/47983071/105987668-5a792780-607d-11eb-9d74-6bd190c0aad2.jpeg" alt="drawing" width="200"/>


This is a Fork from [this package](https://pub.dev/packages/tutorial)

Use the tutorial package is easy.

First create the global keys and add for the components you want to present:
```
final keyMenu = GlobalKey();
final keyContainer = GlobalKey();
final keyChat = GlobalKey();  
```

Then create a list of TutorialItens

```
List<TutorialItens> itens = [];
@override
  void initState() {
    itens.addAll({
      TutorialItens(
          globalKey: keyMenu,
          touchScreen: true,
          top: 200,
          left: 50,
          children: [
            Text(
              "Ali é nosso menu , você consegue ver varias coisas nele",
              style: TextStyle(color: Colors.white, fontSize: 20),
            ),
            SizedBox(
              height: 100,
            )
          ],
          widgetNext: Text(
            "Toque para continuar",
            style: TextStyle(
              color: Colors.purple,
              fontWeight: FontWeight.bold,
            ),
          ),
          shapeFocus: ShapeFocus.oval),
      TutorialItens(
        globalKey: keyChat,
        touchScreen: true,
        top: 200,
        left: 50,
        children: [
          Text(
            "Qualquer duvida que aparecer , entre no nosso chat , estamos prontos para ajudar",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
          SizedBox(
            height: 100,
          )
        ],
        widgetNext: Text(
          "Toque para continuar",
          style: TextStyle(
            color: Colors.purple,
            fontWeight: FontWeight.bold,
          ),
        ),
        shapeFocus: ShapeFocus.oval,
      ),
      TutorialItens(
        globalKey: keyContainer,
        touchScreen: true,
        bottom: 50,
        left: 50,
        children: [
          Text(
            "Nessa sessão você vai ter acesso a todas as  Rasteirinhas",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
          SizedBox(
            height: 10,
          )
        ],
        widgetNext: Text(
          "Sair",
          style: TextStyle(
            color: Colors.purple,
            fontWeight: FontWeight.bold,
          ),
        ),
        shapeFocus: ShapeFocus.square,
      ),
    });
    super.initState();
  }
```

 #### In it we have TutorialItens, is where is the entire tutorial part of your application.
**TutorialItens**
  - **globalKey**  - The Global Key of the component you want to focus on
  - **touchScreen** - Sets whether to move to the next tutorial items by clicking anywhere on the screen
  - **children** - List of widgets to compose the screen
  - **widgetNext** - A component to move to the next item, if touchScreen is equal to false, the click will be only on that component
  - **shapeFocus** -Focus shape can be chosen using **shapeFocus.oval** or **ShapeFocus.square**
  - Can be aligned on the screen as well as positioned using **(top, bottom, left, right)**

**Tutorial.show(context,itens)**
show() -   the show () method, receives two parameters, the context and the list of TutorialItens you created
Now just run the Tutorial.show(context, itens)

```
Tutorial.showTutorial(context, itens);
```
 


CHANGELOG.md



## [0.0.1] - TODO: Add release date.

* First commit.

## [0.0.2] - Simplifying widget.

* Simplifying widget.

## [0.0.3] - Enabling touch function on the screen.

* Enabling touch function on the screen.

## [0.0.4] - changing component size.

* Enabling touch function on the screen.

## [0.0.5] - Readme.

* Arrumando exemplo e readme

## [0.0.6] - Readme.

* Arrumando exemplo e readme

## [0.0.7] - Readme.

* Arrumando exemplo e readme

## [0.0.8] - Readme.

* Migrado para Null Safety

## [1.0.9] - Readme.

* Correção de Null Safety.

## [1.1.0] - Readme.

* Added Implementation of Buttons for Navigation inside tutorial

## [1.1.1] - Readme.

* Fixed Layout of ButtonBar


lib/src/painter/painter.dart
import 'package:flutter/material.dart';
import 'package:navigable_tutorial/src/models/shape_models.dart';

class HolePainter extends CustomPainter {
  final double? dx;
  final double? dy;
  final double? width;
  final double? height;
  final ShapeFocus? shapeFocus;

  HolePainter({
    this.dx,
    this.dy,
    this.width,
    this.height,
    this.shapeFocus = ShapeFocus.oval,
  });
  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()..color = Color.fromRGBO(0, 0, 0, 0.8);
    if (shapeFocus == ShapeFocus.oval) {
      canvas.drawPath(
          Path.combine(
            PathOperation.difference,
            Path()..addRect(Rect.fromLTWH(0, 0, size.width, size.height)),
            Path()
              ..addOval(
                  Rect.fromCircle(center: Offset(dx!, dy!), radius: width!))
              ..close(),
          ),
          paint);
    } else {
      canvas.drawPath(
          Path.combine(
            PathOperation.difference,
            Path()..addRect(Rect.fromLTWH(0, 0, size.width, size.height)),
            Path()
              ..addRect(Rect.fromLTWH(
                  dx! - (width! / 2), dy! - (height! / 2), width!, height!))
              ..close(),
          ),
          paint);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }
}


lib/src/models/tutorial_items.dart
import 'package:flutter/material.dart';
import 'package:navigable_tutorial/src/models/shape_models.dart';

class TutorialItem {
  final GlobalKey? globalKey;
  final ShapeFocus? shapeFocus;
  final List<Widget>? children;
  final double? top;
  final double? bottom;
  final double? left;
  final double? right;
  final double? buttonBarTop;
  final double? buttonBarBottom;
  final double? buttonBarLeft;
  final double? buttonBarRight;
  final Widget? widgetPrev;
  final Widget? widgetNext;
  final CrossAxisAlignment crossAxisAlignment;
  final MainAxisAlignment mainAxisAlignment;
  final bool? touchScreen;
  final ButtonStyle? buttonStyle;

  TutorialItem({
    this.top,
    this.bottom,
    this.left,
    this.right,
    this.buttonBarTop,
    this.buttonBarBottom,
    this.buttonBarLeft,
    this.buttonBarRight,
    this.globalKey,
    this.children,
    this.shapeFocus,
    this.crossAxisAlignment = CrossAxisAlignment.center,
    this.mainAxisAlignment = MainAxisAlignment.center,
    this.widgetPrev,
    this.widgetNext,
    this.touchScreen = false,
    this.buttonStyle,
  });
}


lib/src/controller/tutorial_tutorial.dart
library tutorial;

import 'package:flutter/material.dart';
import 'package:navigable_tutorial/src/models/tutorial_items.dart';
import 'package:navigable_tutorial/src/painter/painter.dart';

class Tutorial {
  static showTutorial(BuildContext context, List<TutorialItem> children) async {
    int count = 0;
    var size = MediaQuery.of(context).size;
    OverlayState? overlayState = Overlay.of(context);
    List<OverlayEntry> entrys = [];
    children.forEach((element) async {
      var offset = _capturePositionWidget(element.globalKey);
      var sizeWidget = _getSizeWidget(element.globalKey);
      entrys.add(
        OverlayEntry(
          builder: (context) {
            return GestureDetector(
              onTap: element.touchScreen == true
                  ? () {
                      entrys[count].remove();
                      count++;
                      if (count != entrys.length) {
                        overlayState.insert(entrys[count]);
                      }
                    }
                  : () {},
              child: Scaffold(
                backgroundColor: Colors.transparent,
                body: Stack(
                  children: [
                    CustomPaint(
                      size: size,
                      painter: HolePainter(
                          shapeFocus: element.shapeFocus,
                          dx: offset.dx + (sizeWidget.width / 2),
                          dy: offset.dy + (sizeWidget.height / 2),
                          width: sizeWidget.width,
                          height: sizeWidget.height),
                    ),
                    Positioned(
                      top: element.top,
                      bottom: element.bottom,
                      left: element.left,
                      right: element.right,
                      child: Container(
                        width: size.width * 0.8,
                        child: Column(
                          crossAxisAlignment: element.crossAxisAlignment,
                          mainAxisAlignment: element.mainAxisAlignment,
                          children: [
                            ...element.children!,
                          ],
                        ),
                      ),
                    ),
                    Positioned(
                      top: element.buttonBarTop ?? size.height / 1.1,
                      bottom: element.buttonBarBottom,
                      left: element.buttonBarLeft,
                      right: element.buttonBarRight,
                      child: Container(
                        width: size.width,
                        child: ButtonBar(
                          alignment: MainAxisAlignment.spaceBetween,
                          layoutBehavior: ButtonBarLayoutBehavior.padded,
                          children: [
                            ElevatedButton(
                                onPressed: () {
                                  entrys[count].remove();
                                  count--;
                                  if (count >= 0) {
                                    overlayState.insert(entrys[count]);
                                  }
                                },
                                child: element.widgetPrev,
                                style: element.buttonStyle),
                            ElevatedButton(
                                onPressed: () {
                                  entrys[count].remove();
                                  count++;
                                  if (count != entrys.length) {
                                    overlayState.insert(entrys[count]);
                                  }
                                },
                                child: element.widgetNext,
                                style: element.buttonStyle)
                          ],
                        ),
                      ),
                    )
                  ],
                ),
              ),
            );
          },
        ),
      );
    });

    overlayState.insert(entrys[0]);
  }

  static Offset _capturePositionWidget(GlobalKey? key) {
    RenderBox? renderPosition =
        key!.currentContext!.findRenderObject() as RenderBox;

    return renderPosition.localToGlobal(Offset.zero);
  }

  static Size _getSizeWidget(GlobalKey? key) {
    RenderBox? renderSize =
        key!.currentContext?.findRenderObject() as RenderBox;
    return renderSize.size;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:tutorial/navigable_tutorial.dart';

main() {
  runApp(MaterialApp(
    debugShowCheckedModeBanner: false,
    theme: ThemeData(
      iconTheme: IconThemeData(
        color: Colors.black,
      ),
    ),
    home: Home(),
  ));
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  List list = [
    {"url": "assets/images/sandalia.jpg", "x": 2, "y": 2},
    {"url": "assets/images/saltos.jpg", "x": 1, "y": 1},
    {"url": "assets/images/rasteirinha.jpg", "x": 1, "y": 1}
  ];

//Crie suas global keys e adicione aos componentes que deverão ser
//localizados e exibidos.
  var keyMenu = GlobalKey();
  var keyContainer = GlobalKey();
  var keyChat = GlobalKey();

  List<TutorialItem> itens = [];

  //Iniciando o estado.
  @override
  void initState() {
    itens.addAll({
      TutorialItem(
          globalKey: keyMenu,
          touchScreen: true,
          top: 200,
          left: 50,
          children: [
            Text(
              "Ali é nosso menu , você consegue ver varias coisas nele",
              style: TextStyle(color: Colors.white, fontSize: 20),
            ),
            SizedBox(
              height: 100,
            )
          ],
          widgetNext: Text(
            "Toque para continuar",
            style: TextStyle(
              color: Colors.purple,
              fontWeight: FontWeight.bold,
            ),
          ),
          shapeFocus: ShapeFocus.oval),
      TutorialItem(
        globalKey: keyChat,
        touchScreen: true,
        top: 200,
        left: 50,
        children: [
          Text(
            "Qualquer duvida que aparecer , entre no nosso chat , estamos prontos para ajudar",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
          SizedBox(
            height: 100,
          )
        ],
        widgetNext: Text(
          "Toque para continuar",
          style: TextStyle(
            color: Colors.purple,
            fontWeight: FontWeight.bold,
          ),
        ),
        shapeFocus: ShapeFocus.oval,
      ),
      TutorialItem(
        globalKey: keyContainer,
        touchScreen: true,
        bottom: 50,
        left: 50,
        children: [
          Text(
            "Nessa sessão você vai ter acesso a todas as  Rasteirinhas",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
          SizedBox(
            height: 10,
          )
        ],
        widgetNext: Text(
          "Sair",
          style: TextStyle(
            color: Colors.purple,
            fontWeight: FontWeight.bold,
          ),
        ),
        shapeFocus: ShapeFocus.square,
      ),
    });

    ///FUNÇÃO QUE EXIBE O TUTORIAL.
    Future.delayed(Duration(microseconds: 200)).then((value) {
      Tutorial.showTutorial(context, itens);
    });

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: Icon(
          Icons.menu,
          key: keyMenu,
        ),
        elevation: 0,
        iconTheme: IconTheme.of(context),
        centerTitle: true,
        title: Text(
          "DAKITA",
          style: TextStyle(
            color: Colors.black,
          ),
        ),
        backgroundColor: Colors.white,
        actions: [
          Container(
            key: keyChat,
            margin: EdgeInsets.symmetric(horizontal: 5),
            width: 40,
            height: 40,
            decoration: BoxDecoration(
              color: Colors.black,
              shape: BoxShape.circle,
            ),
            child: Center(
              child: Icon(
                Icons.chat_rounded,
                color: Colors.white,
              ),
            ),
          )
        ],
      ),
      body: Container(),
      drawer: Drawer(),
    );
  }
}

class Background extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(
          gradient: LinearGradient(
        colors: [
          Color.fromRGBO(44, 191, 224, 1),
          Color.fromRGBO(129, 71, 243, 1),
        ],
      )),
    );
  }
}
